### PR TITLE
Checking that all nodes in the cluster are >= 8.8.0 before installing…

### LINKED ISSUE #95766: behavioral_analytics-events-final_pipeline parsing fails in a mixed cluster
### Elasticsearch Version

main

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

macOS

### Problem Description

The pipeline defined in `behavioral_analytics-events-final_pipeline.json` uses the new `ignore_missing` setting for the `uri_parts` ingest processor added in #95068.

The pipeline is referenced in the index template `behavioral_analytics-events-settings.json`. 

```
"final_pipeline": "behavioral_analytics-events-final_pipeline",
...
```

That template may be PUT as a node is upgraded during a rolling upgrade, in that mixed cluster scenario only the upgraded nodes recognise the new `ignore_missing` setting. 

### Steps to Reproduce

The problem appears in the rolling upgrade test failure tracked in #95360

```
./gradlew ':x-pack:qa:rolling-upgrade:v8.2.0#twoThirdsUpgradedTest' -Dtests.class="org.elasticsearch.upgrades.MlTrainedModelsUpgradeIT" -Dtests.method="testTrainedModelInference" -Dtests.seed=D0F64DC8162BB439 -Dtests.bwc=true -Dtests.locale=be-BY -Dtests.timezone=Africa/Lubumbashi -Druntime.java=20
```

### Logs (if relevant)

```
[2023-05-02T19:25:31,393][WARN ][o.e.i.IngestService      ] [v8.0.1-2] failed to update ingest pipelines
org.elasticsearch.ElasticsearchParseException: processor [uri_parts] doesn't support one or more provided configuration parameters [ignore_missing]
	at org.elasticsearch.ingest.ConfigurationUtils.readProcessor(ConfigurationUtils.java:588) ~[elasticsearch-8.0.1.jar:8.0.1]
	at org.elasticsearch.ingest.ConfigurationUtils.readProcessor(ConfigurationUtils.java:547) ~[elasticsearch-8.0.1.jar:8.0.1]
	at org.elasticsearch.ingest.ConfigurationUtils.readProcessorConfigs(ConfigurationUtils.java:467) ~[elasticsearch-8.0.1.jar:8.0.1]
	at org.elasticsearch.ingest.Pipeline.create(Pipeline.java:82) ~[elasticsearch-8.0.1.jar:8.0.1]
	at org.elasticsearch.ingest.IngestService.innerUpdatePipelines(IngestService.java:922) ~[elasticsearch-8.0.1.jar:8.0.1]
	at org.elasticsearch.ingest.IngestService.applyClusterState(IngestService.java:898) [elasticsearch-8.0.1.jar:8.0.1]
	at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateAppliers(ClusterApplierService.java:544) [elasticsearch-8.0.1.jar:8.0.1]
	at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateAppliers(ClusterApplierService.java:531) [elasticsearch-8.0.1.jar:8.0.1]
	at org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:503) [elasticsearch-8.0.1.jar:8.0.1]
	at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:428) [elasticsearch-8.0.1.jar:8.0.1]
	at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:154) [elasticsearch-8.0.1.jar:8.0.1]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:717) [elasticsearch-8.0.1.jar:8.0.1]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:260) [elasticsearch-8.0.1.jar:8.0.1]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:223) [elasticsearch-8.0.1.jar:8.0.1]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
	at java.lang.Thread.run(Thread.java:833) [?:?]
	Suppressed: org.elasticsearch.ElasticsearchParseException: processor [uri_parts] doesn't support one or more provided configuration parameters [ignore_missing]
		at org.elasticsearch.ingest.ConfigurationUtils.readProcessor(ConfigurationUtils.java:588) ~[elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.ingest.ConfigurationUtils.readProcessor(ConfigurationUtils.java:547) ~[elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.ingest.ConfigurationUtils.readProcessorConfigs(ConfigurationUtils.java:467) ~[elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.ingest.Pipeline.create(Pipeline.java:82) ~[elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.ingest.IngestService.innerUpdatePipelines(IngestService.java:922) ~[elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.ingest.IngestService.applyClusterState(IngestService.java:898) [elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateAppliers(ClusterApplierService.java:544) [elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateAppliers(ClusterApplierService.java:531) [elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:503) [elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:428) [elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:154) [elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:717) [elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:260) [elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:223) [elasticsearch-8.0.1.jar:8.0.1]
		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
		at java.lang.Thread.run(Thread.java:833) [?:?]
	Suppressed: org.elasticsearch.ElasticsearchParseException: processor [uri_parts] doesn't support one or more provided configuration parameters [ignore_missing]
		at org.elasticsearch.ingest.ConfigurationUtils.readProcessor(ConfigurationUtils.java:588) ~[elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.ingest.common.ForEachProcessor$Factory.create(ForEachProcessor.java:187) ~[?:?]
		at org.elasticsearch.ingest.common.ForEachProcessor$Factory.create(ForEachProcessor.java:168) ~[?:?]
		at org.elasticsearch.ingest.ConfigurationUtils.readProcessor(ConfigurationUtils.java:583) ~[elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.ingest.ConfigurationUtils.readProcessor(ConfigurationUtils.java:547) ~[elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.ingest.ConfigurationUtils.readProcessorConfigs(ConfigurationUtils.java:467) ~[elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.ingest.Pipeline.create(Pipeline.java:82) ~[elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.ingest.IngestService.innerUpdatePipelines(IngestService.java:922) ~[elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.ingest.IngestService.applyClusterState(IngestService.java:898) [elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateAppliers(ClusterApplierService.java:544) [elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateAppliers(ClusterApplierService.java:531) [elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:503) [elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:428) [elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:154) [elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:717) [elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:260) [elasticsearch-8.0.1.jar:8.0.1]
		at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:223) [elasticsearch-8.0.1.jar:8.0.1]
		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
		at java.lang.Thread.run(Thread.java:833) [?:?]
```

### PR DESCRIPTION
**Pull request by afoucret**
Closes #95766

We now avoid installing pipelines before all the nodes in the cluster are satisfying a minimum condition version.

### CODE REVIEW

**Review by kderusso — APPROVED** [review]

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/ingest/PipelineRegistry.java [hunk 1]
```diff
@@ -85,9 +86,18 @@ protected boolean isClusterReady(ClusterChangedEvent event) {
             return false;
         }
 
+        if (getMinClusterVersion() != null && event.state().nodes().getMinNodeVersion().onOrBefore(getMinClusterVersion())) {
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/ingest/PipelineRegistry.java:89]: We do want this to happen in 8.8.0, right? Shouldn't this read like:

```suggestion
        if (getMinClusterVersion() != null && event.state().nodes().getMinNodeVersion().before(getMinClusterVersion())) {
```

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/ingest/PipelineRegistry.java [hunk 2]
```diff
@@ -85,9 +86,18 @@ protected boolean isClusterReady(ClusterChangedEvent event) {
...
+            // Not all node satisfies the minimum version requirement: skipping install.
+            return false;
+        }
+
         return true;
     }
 
+    protected Version getMinClusterVersion() {
```
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/ingest/PipelineRegistry.java:97]: Let's make it an abstract function to ensure that it is not an after thought? Can you also add a comment to describe why it's needed?

Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/ingest/PipelineRegistry.java:97]: The name is also a bit misleading, maybe `getMinSupportedNodeVersion`?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/ingest/PipelineRegistry.java (hunk 1 above)
Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/ingest/PipelineRegistry.java:89]: I find it easier to read with:

```
Version minNodeVersion = event.state().nodes().getMinNodeVersion();
if (getMinClusterVersion().after(minNodeVersion) {
  return false;
}
...
```

**Review by jimczi — APPROVED** [review]
LGTM

**Pull request merged**